akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventConsumerServiceImpl.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.grpc.internal
+
+import akka.Done
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.scaladsl.AskPattern.Askable
+import akka.actor.typed.scaladsl.AskPattern.schedulerFromActorSystem
+import akka.dispatch.ExecutionContexts
+import akka.persistence.EventWriter
+import akka.persistence.query.typed.EventEnvelope
+import akka.projection.grpc.internal.proto.Event
+import akka.projection.grpc.internal.proto.EventConsumerService
+import akka.util.ByteString
+import akka.util.Timeout
+import com.google.protobuf.empty.Empty
+import org.slf4j.LoggerFactory
+
+import java.net.URLEncoder
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+/**
+ * INTERNAL API
+ *
+ * gRPC push protocol service for the consuming side
+ */
+private[akka] object EventConsumerServiceImpl {
+
+  def directJournalConsumer(journalPluginId: String)(implicit system: ActorSystem[_]): EventConsumerServiceImpl = {
+    val eventWriter = system.systemActorOf(
+      EventWriter(journalPluginId),
+      s"EventWriter-${URLEncoder.encode(journalPluginId, ByteString.UTF_8)}")
+    implicit val timeout: Timeout = 5.seconds // FIXME from config or can we get rid of it
+
+    val writeEventToJournal: EventEnvelope[Any] => Future[Done] = envelope =>
+      eventWriter.askWithStatus[Done](replyTo => EventWriter.Write(envelope, replyTo))
+
+    new EventConsumerServiceImpl(writeEventToJournal)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final class EventConsumerServiceImpl(consumeEvent: EventEnvelope[Any] => Future[Done])(
+    implicit system: ActorSystem[_])
+    extends EventConsumerService {
+  private val logger = LoggerFactory.getLogger(classOf[EventConsumerServiceImpl])
+
+  private val protoAnySerialization = new ProtoAnySerialization(system)
+  override def consumeEvent(in: Event): Future[Empty] = {
+    // FIXME do we need to make sure events for the same pid are ordered? Should be single writer per pid but?
```

Comment:
That is taken care of by the Projection on the producer side, but we may receive duplicates.

Reply:
There could be a case where a few events are still in flight on the consumer side, when the connection breaks. Producer reconnects (to another consumer node) and pushes same or more events in parallel with the events that are still in flight and might still be processed.

Anyway, as long as we have Projection that is reading those events on the consumer side I think we are good, since it has ways to detect seqNr gaps and such.

Reply:
I think we'll need a highest-seq-nr read in the writer for this, so far I relied on the writer failing and the remote projection restarting, but that is slow and also stops up all writes for all pids on duplicate which isn't good.

Reply:
I was probably wrong about that, the individual write would fail for a duplicate, restarting only that producer stream, it will not crash the EventWriter itself.